src/core/server/plugins/discovery/plugins_discovery.ts

```diff
@@ -130,27 +124,31 @@ function processPaths$(paths: string[], log: Logger) {
  * directory path and produces an error result if it fails to do so or plugin manifest
  * isn't valid.
  * @param path Path to the plugin directory where manifest should be loaded from.
- * @param packageInfo Kibana package info.
  * @param log Plugin discovery logger instance.
+ * @param core Kibana core values.
  */
-function createPlugin$(
-  path: string,
-  packageInfo: PackageInfo,
-  log: Logger
-): Observable<DiscoveryResult> {
-  return from(parseManifest(path, packageInfo)).pipe(
+function createPlugin$(path: string, log: Logger, core: KibanaCore) {
+  return from(parseManifest(path, core.env.packageInfo)).pipe(
     map(manifest => {
+      const pluginDefinition = require(path);
```

Comment:
What are your concerns with using a class syntax for plugins?

I think all but the most trivial of plugins are going to have state that they need to manage across lifecycle handlers. Subscribing to an observable, for example, essentially always needs to have a corresponding unsubscribe logic in the stop functions, so we need to track that subscription internally. I think classes work well for this.

Reply:
Plugin author can still use classes internally like this:
```ts
class Plugin {
  start() {},
  stop() {},
}

export const plugin = (core, dependencies) => {
  return new Plugin(core, dependencies);
};
```

And it's also easy to manage state even without classes too:
```ts
export const plugin = (core, dependencies) => {
  const state = .....;
  return {
    start() {},
    stop() {},
  };
};
```

I meant that I don't really want to support classes as plugin initializers like we had in `new-platform` branch:
```ts
export const plugin = class Plugin {
  constructor(core, dependencies) {},
  start() {},
  stop() {},
}
```
We were detecting that this is a class and where calling `new` for this class. That sounds unnecessary complex to me... We can always add support for this later though if we think it makes sense.

Reply:
One output of this work should be a single definitive convention for people to define plugins. It's confusing if the kibana repo/official docs employ a bunch of different conventions, and it would be similarly confusing for new developers if third party tutorials/docs varied how plugins were defined.

Factory functions seem inconsistent to me here - we use classes throughout core including to define the entry point to core itself, so I'm not sure why we'd adopt a totally different interface for defining a plugin. I think of plugins as configurable systems, and all the systems in core are classes.

nit: You can export the class directly without a need to store it in a variable, which is a little less confusing imo. Plus, we'd just make the export be the default one way or another, right?

```js
export default class Plugin {
}
```

> We were detecting that this is a class and where calling new for this class.

Why would we need to detect if it's a class? If we say plugins have to export a class, then we can just `new` what they export. If that errors, the plugin is broken. I'd put this in the same category as something like not exporting anything at all.

Unrelated to classes vs factory functions, `core` and `dependencies` will need to be passed into each of the lifecycle hooks because you'll get access to different contracts depending on which part of the lifecycle it is. For example, you won't have access to something like `http.createRouter` inside `plugin.stop`.

The idea is that lifecycle events exposed a client for a system that was tailored to that part of the lifecycle, and then the plugin service would expose those clients to plugins only in the corresponding part of their lifecycle. For example:

```js
class HttpService {
  start() {
    return { createRouter() {} };
  }
}

class CoreSystem {
  start() {
    const http = this.http.start();
    this.plugins.start({ http });
  }
}

class MyPlugin {
  start(core) {
    core.createRouter();
  }
  stop(core) {
    core.createRouter(); // throws
  }
}
```

The convention of a lifecycle handler _returning_ its public contract is the only one that has persisted since the early days of new platform discussions, but since we're in the early days of rolling it out, I'm open to alternative proposals if you have any. Though it's worth noting that this pattern is already in use between core services in the browser system, e.g.: https://github.com/elastic/kibana/blob/master/src/core/public/core_system.ts#L103-L104

Reply:
@azasypkin synced outside of github, and we landed here:

We won't support configurable plugin paths in new platform.

We'll start with a convention where plugins are declared as a class, but the actual integration point with the plugin service itself will be a factory function that instantiates that class. The factory function will not be exposed as a default export.

We will pass unique contracts from core (and plugins) to lifecycle functions, so you'll get a different set of capabilities in `stop` than you will in `start`.

He walked me through the key parts of this PR that I had questions and concerns with as well, and I'm satisfied with the approach.